### PR TITLE
I2C API Enforce stop condition after last message in `i2c_transfer` wrapper

### PR DESCRIPTION
**Pull request by bjarki-andreasen**
The I2C API specifies that the last message of a transfer implies a STOP condition. This requirement has been offloaded to each device driver to implement, which results in many in-tree drivers (more than half of them) not following this requirement.

Since the STOP is implied by the API, we can ensure the STOP bit of the last message is set before passing the messages to the device drivers for transfer. This ensures all device drivers implicitly follow this requirement, allowing us to remove this now duplicate code from the in-tree device drivers which currently do this internally.

We can additionally prevent calls to the `transfer` and `transfer_cb` APIs if `num_msgs` is 0, which is a check we need to do anyway before setting the last messages STOP bit, allowing us to return early if there is nothing to transmit, as is described by the API as well: "@a num_msgs may be zero,in which case no transfer occurs."

### CODE REVIEW

File: include/zephyr/drivers/i2c.h [hunk 1]
```diff
@@ -531,6 +531,15 @@ static inline void i2c_dump_msgs(const struct device *dev, const struct i2c_msg
...
 }
 
+/** @cond INTERNAL_HIDDEN */
+
+static inline void i2c_set_last_msg_stop(struct i2c_msg *msgs, uint8_t num_msgs)
+{
+	msgs[num_msgs - 1].flags |= I2C_MSG_STOP;
+}
+
+/** @endcond */
```
Comment (gmarull) [inline include/zephyr/drivers/i2c.h:541]: is this inline helper really necessary?

Comment (bjarki-andreasen) [reply include/zephyr/drivers/i2c.h:541]: no, but it is pretty :)

Comment (ydamigos) [reply include/zephyr/drivers/i2c.h:541]: `msgs[num_msgs - 1].flags |= I2C_MSG_STOP;` is straight forward.
I don't see the point of adding 8 lines of code if there is only one call to the inline function.

Comment (bjarki-andreasen) [reply include/zephyr/drivers/i2c.h]: Understood :)

**Comment by bjarki-andreasen** [comment]
Removed the wrapper function :)

**Review by ydamigos — DISMISSED** [review]

File: include/zephyr/drivers/i2c.h [hunk 2]
```diff
@@ -785,6 +785,8 @@ static inline int z_impl_i2c_transfer(const struct device *dev,
 	const struct i2c_driver_api *api =
 		(const struct i2c_driver_api *)dev->api;
 
+	msgs[num_msgs - 1].flags |= I2C_MSG_STOP;
```
Comment (teburd) [inline include/zephyr/drivers/i2c.h:788]: Please add an assert before this verifying num_msgs >= 1

**Review by teburd — DISMISSED** [review]

File: include/zephyr/drivers/i2c.h (hunk 2 above)
Comment (bjarki-andreasen) [reply include/zephyr/drivers/i2c.h:788]: ok, did not notice we allow a length of zero :) Out of curiosity, why do we allow this?

**Review by uLipe — DISMISSED** [review]

**Review by ydamigos — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]

**Review by soburi — DISMISSED** [review]

**Comment by bjarki-andreasen** [comment]
I have updated the wrapper to check if `num_msgs == 0` and immediately return instead of invoking the `transfer` implementation given that no transaction will occur, so there is no reason to call it. This further optimized device drivers since they don't have to do this check internally.

Additionally, I implemented the set stop condition + check for `num_msgs == 0` to the async version of `i2c_transfer`, `i2c_transfer_cb` which I did not notice initially :)

**Review by erwango — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Review by ydamigos — APPROVED** [review]

**Pull request merged**